### ISSUE TITLE
--opencl-devices parameter not working

### ISSUE THREAD

**Issue by ScottPierce** [ethereum-mining/ethminer#152]
I've tried calling `ethminer.exe --list-devices`, and then using those device numbers with the `--opencl-devices` param to skip my integrated graphics, but it doesn't seem to be working. It still seems to select my integrated graphics card and then it tells me that my graphics card doesn't have enough memory to upload the dag file.

**Comment by kiwina** [comment]
can confirm the same error even picking one device dosn't work

**Comment by roninkaizen** [comment]
looked it up while testing and works fine for me, 
driver?OS?

**Comment by ScottPierce** [comment]
Windows 10, and I'm using the latest driver for 8x gtx 1070. I found this
issue when I only had 4x cards though. My end goal is to simply stop
ethminer from trying to use the integrated graphics.

**Comment by roninkaizen** [comment]
@ScottPierce - then define it by stating which device should be used like "opencl-device 0 1 2 3" or the correspondig cuda-device
-comfortable as you think, one of the gpus is erronous
feel free to chat me in gitter

**Comment by Rootax** [comment]
Same here, --opencl-devices have no effect, it's always selecting 0, even if I try to force 1.

**Comment by michael-pesce** [comment]
--opencl-devices is not the right parameter for NVIDIA assuming you're running with -U (CUDA).

Instead use --cuda-devices

**Comment by Rootax** [comment]
I use amd cards, in opencl mode.

**Comment by michael-pesce** [comment]
Was looking at OP's issue. 

I always disable integrated graphics in BIOS. I know that will fix the problem! :)

**Comment by GSI** [comment]
Can confirm the issue in both 0.11.0 and 0.12.0.dev2

Linux with two AMD-based cards (no onboard graphics); `--list-devices` shows both, but `--opencl-device` and `--opencl-devices` both seem to be simply without effect.

I do not know what `--opencl-platform` is about, but setting it to **1** causes **ethminer** to quit immediately.

**Comment by mxjoe** [comment]
I have the same problem under macOS 10.12.6 with latest ethminer 0.12.0dev2. Ethminer always uses the integrated Intel GPU instead of the Radeon Pro 580 I told ethminer to use.

**Comment by madlink** [comment]
any news on this issue? i still having the same problem

**Comment by smurfy** [comment]
please post your output of ``--list-devices``

**Comment by jamesatfish** [comment]
I have the same issue trying to run ethminer on Mac OS 10.13.2 on a late 2017 MacBook Pro.

--list-devices shows both Intel and AMD graphics cards:

```
Listing OpenCL devices.
FORMAT: [platformID] [deviceID] deviceName
[0] [0] Intel(R) HD Graphics 630
	CL_DEVICE_TYPE: GPU
	CL_DEVICE_GLOBAL_MEM_SIZE: 1610612736
	CL_DEVICE_MAX_MEM_ALLOC_SIZE: 402653184
	CL_DEVICE_MAX_WORK_GROUP_SIZE: 256
[0] [1] AMD Radeon Pro 560 Compute Engine
	CL_DEVICE_TYPE: GPU
	CL_DEVICE_GLOBAL_MEM_SIZE: 4294967296
	CL_DEVICE_MAX_MEM_ALLOC_SIZE: 1073741824
	CL_DEVICE_MAX_WORK_GROUP_SIZE: 256
```

Despite this, no matter what deviceID I specify with `--opencl-device x` ethminer selects the Intel graphics card.

**Comment by DeadManWalkingTO** [comment]
OpenCL Hardware Test Launch Command:

`ethminer -RH -G -S eu1.ethermine.org:4444 -FS us1.ethermine.org:4444 -O 0x7013275311fc37ccc1e40193D75086293eCb43A4.issue152`

Try with latest Ethminer version and feedback please.
Thank you!

**Comment by jamesatfish** [comment]
Ran the command using 0.14.0.dev2.

I can see that the AMD 560 GPU is identified and begins to mine after the DAG is created for the card.

Unfortunately it's only mining at less than 2Mh/s when that GPU should be capable of 10-12Mh/s.

```
Volume:bin 3 user$ ./ethminer -RH -G -S eu1.ethermine.org:4444 -FS us1.ethermine.org:4444 -O 0x7013275311fc37ccc1e40193D75086293eCb43A4.issue152
  m  09:49:50|        |  ethminer version 0.14.0.dev2
  m  09:49:50|        |  Build: Darwin/unknown / Release
  ℹ  09:49:50|        |  Found suitable OpenCL device [ Intel(R) HD Graphics 630 ] with 1610612736  bytes of GPU memory
  m  09:49:50|        |  not-connected
  ℹ  09:49:51|stratum |  Connected to eu1.ethermine.org
  ℹ  09:49:51|stratum |  Spinning up miners...
 cl  09:49:51|cl-0    |  No work. Pause for 3 s.
 cl  09:49:51|cl-1    |  No work. Pause for 3 s.
  ℹ  09:49:51|stratum |  Subscribed to stratum server
  ℹ  09:49:52|stratum |  Authorized worker 0x7013275311fc37ccc1e40193D75086293eCb43A4.issue152
  ℹ  09:49:52|stratum |  Received new job #52a89c4a from eu1.ethermine.org
 cl  09:49:54|cl-0    |  New work: header #52a89c4a… target 0000000112e0be826d694b2e62d01511f12a6061fbaec8bc02357593e70e52ba
 cl  09:49:54|cl-1    |  New work: header #52a89c4a… target 0000000112e0be826d694b2e62d01511f12a6061fbaec8bc02357593e70e52ba
 cl  09:49:54|cl-0    |  New seed #1261dfe1…
 cl  09:49:54|cl-1    |  New seed #1261dfe1…
 cl  09:49:55|cl-0    |  Platform: Apple
 cl  09:49:55|cl-1    |  Platform: Apple
 cl  09:49:55|cl-0    |  Device:   Intel(R) HD Graphics 630  / OpenCL 1.2 
 cl  09:49:55|cl-1    |  Device:   AMD Radeon Pro 560 Compute Engine  / OpenCL 1.2 
 cl  09:49:55|cl-0    |  OpenCL kernel: Stable kernel
  m  09:49:55|        |  Speed   0.00 Mh/s    gpu/0  0.00  gpu/1  0.00  [A0+0:R0+0:F0] Time: 00:00
 cl  09:49:55|cl-0    |  Build info: <program source>:334:9: warning: comparison of integers of different signs: 'int' and 'const uint' (aka 'const unsigned int')
                if (i == thread_id)
      ~ ^  ~~~~~~~~~
<program source>:369:9: warning: comparison of integers of different signs: 'int' and 'const uint' (aka 'const unsigned int')
                if (i == thread_id)
      ~ ^  ~~~~~~~~~

  ℹ  09:49:55|cl-0    |  OpenCL device Intel(R) HD Graphics 630  has insufficient GPU memory. 1610612736  bytes of memory found < 2516582272  bytes of memory required
  ✘  09:49:55|cl-0    |  OpenCL Error: clEnqueueWriteBuffer: CL_INVALID_MEM_OBJECT (-38)
 cl  09:49:56|cl-1    |  OpenCL kernel: Stable kernel
 cl  09:49:56|cl-1    |  Build info: <program source>:336:9: warning: comparison of integers of different signs: 'int' and 'const uint' (aka 'const unsigned int')
                if (i == thread_id)
      ~ ^  ~~~~~~~~~
<program source>:371:9: warning: comparison of integers of different signs: 'int' and 'const uint' (aka 'const unsigned int')
                if (i == thread_id)
      ~ ^  ~~~~~~~~~

 cl  09:49:56|cl-1    |  Creating light cache buffer, size 39320768
 cl  09:49:56|cl-1    |  Creating DAG buffer, size 2516582272
 cl  09:49:56|cl-1    |  Loading kernels
 cl  09:49:56|cl-1    |  Writing light cache buffer
 cl  09:49:56|cl-1    |  Creating buffer for header.
 cl  09:49:56|cl-1    |  Creating mining buffer
  ℹ  09:49:57|stratum |  Received new job #fc84576d from eu1.ethermine.org
  m  09:50:00|        |  Speed   0.00 Mh/s    gpu/0  0.00  gpu/1  0.00  [A0+0:R0+0:F0] Time: 00:00
  ℹ  09:50:01|stratum |  Received new job #f07216af from eu1.ethermine.org
  ℹ  09:50:03|stratum |  Received new job #9ed85f88 from eu1.ethermine.org
  m  09:50:05|        |  Speed   0.00 Mh/s    gpu/0  0.00  gpu/1  0.00  [A0+0:R0+0:F0] Time: 00:00
  m  09:50:10|        |  Speed   0.00 Mh/s    gpu/0  0.00  gpu/1  0.00  [A0+0:R0+0:F0] Time: 00:00
  ℹ  09:50:11|cl-1    |  2.34375  GB of DAG data generated in 14985 ms.
 cl  09:50:11|cl-1    |  Switch time 8120 ms / 17127448 us
 cl  09:50:11|cl-1    |  New work: header #9ed85f88… target 0000000112e0be826d694b2e62d01511f12a6061fbaec8bc02357593e70e52ba
 cl  09:50:11|cl-1    |  Switch time 8121 ms / 174 us
  ℹ  09:50:13|stratum |  Received new job #743496d1 from eu1.ethermine.org
 cl  09:50:13|cl-1    |  New work: header #743496d1… target 0000000112e0be826d694b2e62d01511f12a6061fbaec8bc02357593e70e52ba
 cl  09:50:13|cl-1    |  Switch time 316 ms / 284 us
  ℹ  09:50:15|stratum |  Received new job #922cef0b from eu1.ethermine.org
 cl  09:50:15|cl-1    |  New work: header #922cef0b… target 0000000112e0be826d694b2e62d01511f12a6061fbaec8bc02357593e70e52ba
 cl  09:50:15|cl-1    |  Switch time 27 ms / 314 us
  m  09:50:15|        |  Speed   1.93 Mh/s    gpu/0  0.00  gpu/1  1.93  [A0+0:R0+0:F0] Time: 00:00
  ℹ  09:50:17|stratum |  Received new job #929aa2a5 from eu1.ethermine.org
 cl  09:50:17|cl-1    |  New work: header #929aa2a5… target 0000000112e0be826d694b2e62d01511f12a6061fbaec8bc02357593e70e52ba
 cl  09:50:17|cl-1    |  Switch time 418 ms / 375 us
  m  09:50:20|        |  Speed   1.84 Mh/s    gpu/0  0.00  gpu/1  1.84  [A0+0:R0+0:F0] Time: 00:00
  m  09:50:25|        |  Speed   1.83 Mh/s    gpu/0  0.00  gpu/1  1.83  [A0+0:R0+0:F0] Time: 00:00
  m  09:50:30|        |  Speed   1.72 Mh/s    gpu/0  0.00  gpu/1  1.72  [A0+0:R0+0:F0] Time: 00:00
  ℹ  09:50:31|stratum |  Received new job #59a41dae from eu1.ethermine.org
 cl  09:50:32|cl-1    |  New work: header #59a41dae… target 0000000112e0be826d694b2e62d01511f12a6061fbaec8bc02357593e70e52ba
 cl  09:50:32|cl-1    |  Switch time 481 ms / 302 us
  m  09:50:35|        |  Speed   1.78 Mh/s    gpu/0  0.00  gpu/1  1.78  [A0+0:R0+0:F0] Time: 00:00
^C  ℹ  09:50:40|        |  Disconnected from eu1.ethermine.org
  ℹ  09:50:40|        |  Shutting down miners...
```

**Comment by DeadManWalkingTO** [comment]
OpenCL Low Hashrate Report
#308

**Comment by sirplus51** [comment]
See my post 
https://github.com/ethereum-mining/ethminer/issues/1306#issuecomment-400416030
My work around is almost at the end
1/limit platforms to 2
2/Mine opencl on the whole platform (don't specify devices or it will try to mine those cuda devices as opencl, detect thats not efficient and then throttle them back)

**Comment by colin-i** [comment]
Same error.

> Listing OpenCL devices.
FORMAT: [platformID] [deviceID] deviceName
[0] [0] P106-100
        CL_DEVICE_TYPE: GPU
        CL_DEVICE_GLOBAL_MEM_SIZE: 6372589568
        CL_DEVICE_MAX_MEM_ALLOC_SIZE: 1593147392
        CL_DEVICE_MAX_WORK_GROUP_SIZE: 1024
[1] [0] Oland
        CL_DEVICE_TYPE: GPU
        CL_DEVICE_GLOBAL_MEM_SIZE: 3221225472
        CL_DEVICE_MAX_MEM_ALLOC_SIZE: 3221225472
        CL_DEVICE_MAX_WORK_GROUP_SIZE: 256
[1] [1] Scrapper
        CL_DEVICE_TYPE: GPU
        CL_DEVICE_GLOBAL_MEM_SIZE: 2147483648
        CL_DEVICE_MAX_MEM_ALLOC_SIZE: 536870912
        CL_DEVICE_MAX_WORK_GROUP_SIZE: 256
Listing CUDA devices.
FORMAT: [deviceID] deviceName
[0] P106-100
        Compute version: 6.1
        cudaDeviceProp::totalGlobalMem: 6372589568
        Pci: 0000:03:00

And .\ethminer -X --opencl-platform 1 --opencl-device 0 -P ...

Is selecting Scrapper with any UINT at --opencl-device

**Comment by AndreaLanfranchi** [comment]
Resolved by #1704
